### PR TITLE
Add server initial metadata test for callback api

### PR DESCRIPTION
**Pull request by hcaseyal**
Adds a test for server initial metadata in the callback api

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] Performance differences noted:
Benchmark                                                                                    locks_per_iteration
-------------------------------------------------------------------------------------------  ---------------------
BM_StreamingPingPong<InProcess, NoOpMutator, NoOpMutator>/134217728/1                        +4%
BM_StreamingPingPong<InProcess, NoOpMutator, NoOpMutator>/134217728/2                        +4%
BM_StreamingPingPong<MinInProcess, NoOpMutator, NoOpMutator>/134217728/1                     +4%
BM_StreamingPingPong<MinInProcess, NoOpMutator, NoOpMutator>/134217728/2                     +4%
BM_StreamingPingPongMsgs<InProcess, NoOpMutator, NoOpMutator>/134217728                      +4%
BM_StreamingPingPongMsgs<MinInProcess, NoOpMutator, NoOpMutator>/134217728                   +4%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/134217728/1/0     +4%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/134217728/1/1     +5%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/134217728/2/0     +4%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/134217728/2/1     +4%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/134217728/1/0  +4%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/134217728/1/1  +5%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/134217728/2/0  +4%
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/134217728/134217728                 +5%
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,751      Total (=)      2,019,751

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,154,121      Total (>)     11,154,119

 No significant differences in binary sizes


```

**Comment by hcaseyal** [comment]
Ping

File: test/cpp/end2end/test_service_impl.cc [hunk 1]
```diff
@@ -34,8 +34,22 @@ using std::chrono::system_clock;
 
 namespace grpc {
 namespace testing {
-namespace {
 
+int MetadataMatchCount(
```
Comment (vjpai) [inline test/cpp/end2end/test_service_impl.cc]: If this free function is now being used by both client and server code, it should probably be refactored into a separate file called `utils_common.cc` or something like that in this directory.

File: test/cpp/end2end/test_service_impl.h [hunk 2]
```diff
@@ -38,6 +38,14 @@ const char* const kServerFinishAfterNReads = "server_finish_after_n_reads";
 const char* const kServerUseCoalescingApi = "server_use_coalescing_api";
 const char* const kCheckClientInitialMetadataKey = "custom_client_metadata";
 const char* const kCheckClientInitialMetadataVal = "Value for client metadata";
+const char* const kCheckServerInitialMetadataKey = "custom_server_metadata";
+const char* const kCheckServerInitialMetadataVal = "Value for server metadata";
+
+// Returns the number of pairs in metadata that exactly match the given
+// key-value pair. Returns -1 if the pair wasn't found.
+int MetadataMatchCount(
```
Comment (vjpai) [inline test/cpp/end2end/test_service_impl.h]: Likewise with this function declaration in this header file.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,676      Total (=)      2,019,676

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,154,189      Total (>)     11,154,186

 No significant differences in binary sizes


```

**Comment by hcaseyal** [comment]
@hcaseyal reminder to check on the test failures and try to merge this

**Comment by vjpai** [comment]
@hcaseyal Is this PR still relevant? Are you going to adjust it as needed?

**Comment by hcaseyal** [comment]
@vjpai It's still relevant, but I don't have cycles right now to deal with the merge conflicts

**Pull request closed without merging**